src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts

```diff
@@ -761,10 +761,22 @@ export abstract class AbstractTaskService extends Disposable implements ITaskSer
 		}
 	}
 
-	private setRecentlyUsedTask(task: Task): void {
-		const key = task.getRecentlyUsedKey();
+	private async setRecentlyUsedTask(task: Task): Promise<void> {
+		let key = task.getRecentlyUsedKey();
 		if (!InMemoryTask.is(task) && key) {
-			this.getRecentlyUsedTasks().set(key, JSON.stringify(this.createCustomizableTask(task)));
+			const customizations = this.createCustomizableTask(task);
+			if (ContributedTask.is(task) && customizations) {
+				let custom: CustomTask[] = [];
+				let customized: IStringDictionary<ConfiguringTask> = Object.create(null);
+				await this.computeTasksForSingleConfig(task._source.workspaceFolder ?? this.workspaceFolders[0], {
+					version: '2.0.0',
+					tasks: [customizations]
+				}, TaskRunSource.System, custom, customized, TaskConfig.TaskConfigSource.TasksJson, true);
+				for (const configuration in customized) {
+					key = customized[configuration].getRecentlyUsedKey()!;
```

Comment:
Why is the key a random key from the customized hash table ?

Reply:
The key is from here https://github.com/microsoft/vscode/blob/8918d05cb5b88cb43a45a7d9f04d7d64491c494a/src/vs/workbench/contrib/tasks/common/tasks.ts#L731-L731
I start with a `ContributedTask` key, but I'm storing a `CustomTask`, so when I re-create the task later it will have a `CustomTask` key. To avoid this, I convert the `ContributedTask` into a `CustomTask` then use that key.

Reply:
What I actually mean is the following: how can this piece of code be deterministic given that we loop over the keys in a hash table. There is no order associated with this. Or does it simply not matter.

Reply:
There will be only one key in the hash table, since only one task is fed in. I will make that clearer.

Reply:
OK. That explains it to me. Thanks.